cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp

```diff
@@ -59,6 +61,171 @@ inline sycl::event beta_copy_transform(sycl::queue& queue,
+    sycl::event& event_A,
+    ndview<Float, 2>& b, // t(X)*y, RHS, solution is outputted here
+    sycl::event& event_b,
+    const std::int64_t dim_A,
+    const std::int64_t nrhs) {
+    constexpr auto alloc = sycl::usm::alloc::device;
+
+    /* Decompose: A = Q * diag(l) * t(Q) */
+    /* Note: for NRHS>1, this will overallocate in order to reuse the memory as buffer later on */
+    auto eigenvalues = ndarray<Float, 1>::empty(queue, dim_A * nrhs, alloc);
+    sycl::event syevd_event =
+        syevd<mkl::job::vec, uplo, Float>(queue, dim_A, A, dim_A, eigenvalues, { event_A });
+    const Float eps = std::numeric_limits<Float>::epsilon();
+
+    /* Discard too small singular values */
+    std::int64_t num_discarded;
+    {
+        /* This is placed inside a block because the array created here is not used any further */
+        /* Note: the array for 'eigenvalues' is allocated with size [dimA, nrhs],
+        but by this point, only 'dimA' elements will have been written to it. */
+        auto eigenvalues_cpu =
+            ndview<Float, 1>::wrap(eigenvalues.get_data(), dim_A).to_host(queue, { syevd_event });
+        const Float* eigenvalues_cpu_ptr = eigenvalues_cpu.get_data();
+        const Float largest_ev = eigenvalues_cpu_ptr[dim_A - 1];
+        if (largest_ev <= eps) {
+            throw std::runtime_error(
+                "Could not solve linear system. Problem has too small singular values.");
+        }
+        const Float component_threshold = eps * largest_ev;
+        for (num_discarded = 0; num_discarded < dim_A - 1; num_discarded++) {
+            if (eigenvalues_cpu_ptr[num_discarded] > component_threshold)
+                break;
+        }
+    }
+
+    /* Create the square root of the inverse: Qis = Q * diag(1 / sqrt(l)) */
+    std::int64_t num_taken = dim_A - num_discarded;
+    auto ev_mutable = eigenvalues.get_mutable_data();
+    sycl::event inv_sqrt_eigenvalues_event = queue.submit([&](sycl::handler& h) {
+        h.depends_on(syevd_event);
+        h.parallel_for(num_taken, [=](const auto& i) {
+            const std::size_t ix = i + num_discarded;
+            ev_mutable[ix] = sycl::sqrt(Float(1) / ev_mutable[ix]);
+        });
+    });
+
+    auto Q_mutable = A.get_mutable_data();
+    sycl::event inv_sqrt_eigenvectors_event = queue.submit([&](sycl::handler& h) {
+        const auto range = oneapi::dal::backend::make_range_2d(num_taken, dim_A);
+        h.depends_on(inv_sqrt_eigenvalues_event);
+        h.parallel_for(range, [=](sycl::id<2> id) {
+            const std::size_t col = id[0] + num_discarded;
+            const std::size_t row = id[1];
+            Q_mutable[row + col * dim_A] *= ev_mutable[col];
+        });
+    });
+
+    /* Now calculate the actual solution: Qis * Qis' * B */
+    const std::int64_t eigenvectors_offset = num_discarded * dim_A;
+    if (nrhs == 1) {
+        sycl::event gemv_right_event =
+            mkl::blas::column_major::gemv(queue,
```

Comment:
Please use existing wrappers to oneMKL:
https://github.com/oneapi-src/oneDAL/blob/main/cpp/oneapi/dal/backend/primitives/blas/gemv.hpp

Or add a new wrapper if it is needed.

Reply:
Changed them to the internal wrappers for `gemm` and `gemv`, but I'd say their logic is quite hard to follow.